### DISCUSSION TITLE
Visual offset option, new implementation proposal

### DISCUSSION THREAD

**Discussion by Adrriii** [ppy/osu#26242] [Ideas]
First of all, I'll state that I'm aware ; This is far from being a priority. This can be kept for later and feel free to ignore for now.

In the older discussion #13143, peppy came to the conclusion that : 
> Other games that provide visual offsets only do so because they either
> - don't have hitsounds
> - silence hitsounds when in effect

However, in many **current** situations it is highly preferable to turn off hitsounds and apply a large audio offset (-50ms and below), paired with skin customizations in order to account for the resulting discrepancies. This has a negative impact on many devices support.

For the sake of completeness, I'll quickly state what audio/visual offsets are good for and why both should be fixed.

---

### Audio offset is good for **audio output latency**:
- Bluetooth / Wifi equipment
- Sound processing interfaces
- Overall slow sound I/O

If your setup works correctly otherwise, and since such latency is pretty much constant, you're good to play after simply applying the relevant delay. This can usually be fixed otherwise with cheap alternatives (analog audio, using speakers ...)

### Visual offset is good for **display and input latency**:
- Bad keyboard / firmware
- Bad display device
- Bad user settings but not necessarily aware / too complex
- Overall slow computer I/O / drivers

Much more common and not necessarily constant, this is what may cause most latency issues among players. This is much harder to fix outside of the game, especially if you're low income. Even investing several hundred/thousand dollars in equipment may still leave you behind if you're unlucky (especially gaming laptops).

---

In the previous discussion, a solution was put forward regarding literally shifting the offset of gameplay rendering. I am totally unaware of the complexity of such a feat, but I imagine it is a very complex endeavor without messing with beatmap parsing and such.

It also seemed like it wouldn't be worth it without support for hitsounds, which is literally impossible as stated by peppy. I think it's always worth it to give the opportunity to the player to chose his tradeoff.

What if a much simpler approach could be tried ? Following these steps:
- Add the visual offset option (one-way only)
- When the option is used, HitObjects will disappear as if hit on-time unless they are hit too early (to avoid any complexity with their display)
- User hits are offset by the set value during hit registration
- (optional) Warn the user about hitsounds delay, and the possibility of turning them off

It would also be technically possible to autoplay hitsounds as an option, but maybe it's not desirable to include this.

Please tell me what you think or if you know a better implementation. Thanks for reading.

**Comment by bdach** [comment]
mandatory links for yet another offset discussion
- https://gist.github.com/smoogipoo/b2a9b95b4ea258345aaf599e1b47cbb5
- https://discord.com/channels/188630481301012481/188630652340404224/1086675014402723860

> Following these steps:
> * Add the visual offset option (one-way only)
> * When the option is used, HitObjects will disappear as if hit on-time unless they are hit too early (to avoid any complexity with their display)
> * User hits are offset by the set value during hit registration
> * (optional) Warn the user about hitsounds delay, and the possibility of turning them off

this sounds pretty terrible. so the map should basically half-autoplay itself and the player is supposed to pretend they're the ones hitting the objects?

also hitsounding is a key component of osu! beatmaps. to the degree some maps are literally not complete without them (cf. keysounding). i don't think turning them off should be in any way encouraged.

**Reply by Adrriii** [reply]
> so the map should basically half-autoplay itself and the player is supposed to pretend they're the ones hitting the objects?

In rather low-latency scenarios this would indeed not be a preferrable solution. Maybe a fade with a shorter timespan the stronger the offset is could have better results. This autoplay is merely an attempt at reducing artefacts and may not be necessary after all.

In higher-latency scenarios, it's preferrable to have such rendering artefacts and still be able to hit on time. Especially in scrolling games like taiko and mania. In mania it would resemble the skin workaround solution except it also works with speed changes and different scroll speeds.

> hitsounding is a key component of osu! beatmaps. to the degree some maps are literally not complete without them

I'm aware. Audio offset already makes hitsounds weird. I play with around -50ms. For hitsounded maps, I need to set it to 0ms and enable hitsounds. These maps are very hard to play as-is, but there isn't really another way around. However in mania such maps have not been ranked in, at least, years. I can only think of 2 in my keymode at the top of my head. And a Loved mappack. I think it's not a good thing to keep a broken state for edge-case scenarios that can just be dealt with on the fly. I know mania is just one gamemode, but it's one where visual offset is 10x more relevant than considering hitsounds that are likely to be disabled by the player.

I'd like to point out that, some players mentioned being happy with just being able to set a much higher hitposition (like, out of the screen). Seems like a mich simpler implementation for you. A simple "hit offset" without any other measures could also work fine.

**Reply by OWazabi** [reply]
I think the point of view from players and devs differs a lot here.

I'd highly recommend to anyone wanting to play a VSRGs to disable hit sounds and set up an offset calibration, as they just drive you away from improvement and understanding of the game, while also making your player experience slightly worse by playing over the music's and butchering them.

Ironically, I think there's a huge offset between what players want as gameplay and what is being implemented.
I believe people rather have proper settings, for offsets, skins and whatnot, whatever makes them able to play and progress in a better way, instead of some cute piano sounds when you press a key on a few meme and kawaii charts.

Kindly,
Waza.

**Comment by peppy** [comment]
give me one example where it's beneficial to have two offsets instead of one. you've failed to cover that.

unless you example is "disable hit sounds plz" warning, which arguably doesn't even need a second offset of we were to entertain the idea.

**Reply by Adrriii** [reply]
> give me one example where it's beneficial to have two offsets instead of one. you've failed to cover that.

Sure, audio and visual offset are completely independant in what they achieve. This is detailed in my original message. Here is an example:

Take a setup with a total delay of 70ms (rather common with laptops). A player trying to trouble shoot offset will go in two parts : Audio delay (blind/wizard), Visual delay (hide judges + hide hiterror).

Say that after calibration the offset wizard recommends -30ms : That's the audio offset portion. *Depending on the player, it might be enough to have to turn off hitsounds already due to the negative delay.*

Then playing a map without sound and without judgement feedback will give a solid 40ms late median, which would mean a required +40ms visual offset (or, in a simpler implementation, a -40ms hit offset)

Currently, there are a few workarounds:
- Stay as-is and manually compensate during gameplay. This solution has high cognitive dissonance.
- Apply a -70ms audio offset and try to time only by audio. This requires adaptation and may not work at all for seasoned players with their habits.
- (mania only) "cheat the visual offset" with skinning. *On a set scroll speed only*, you can emulate a visual offset by increasing the hitposition a relevant amount, then skinning the HitReceptors appropriately to be above the judgement line. For best results, hide what's below the HitReceptors. This solution falls apart at the first speed change encounter, works alright in a controlled environment if you ignore LN holds.
- Spend hundreds or thousands in new equipment and hope for low latency. Reach for the community for equipment advice. This solution requires to not be poor.

**Reply by OWazabi** [reply]
It is a great deal honestly, and single handedly one of the best features any other game than o!m has (paired with custom rates, those are the 2 main lacking o!m features)

Adri explanation, while on point, is not entirely accurate imo because he's really technical, when most of the players aren't as much.
Let me explain with my simpler words and situations.

Audio Offset, is def non related to visual offset.
The song has a timing, your computer has a delay, you find that precise delay, set up the offset, and there's never a good reason to change it ever again unless your audio delay actually changed. (New hardware for example)


Visual delay, is the magical "fix your accuracy and distortionned habits" tool.
Whatever is truly bugging you and making you play kinda "off", visual offset is your fixer.

Maybe you've been used to playing 35 scroll speed o!m before understanding that isn't appropriated to your current level, and now if you want to play on 28 for better reading, you're hitting way too early.
Maybe your new keyboard has an bunch of delay, and fixing audio offset doesn't change anything for you since it's really based on the player playstyle.

The point is, it doesn't matter why you need it, maybe it isn't even a legitimate issue you're facing, and could be fixed somehow else.
At the end of the day, it doesn't matter.
Visual Offset is the best QoL tool to fix your actual hits, effectively.

Not having this tool feels like playing in stone age on o!m, when compared to what's being done elsewhere.
 
**The simplicity and efficacy of just saying "ok well, something is off, I've been hitting 10ms late on average for the past few maps"
Then just putting 10ms visual offset delay, and instantly being spot on...
That's true QoL and would help anyone enjoying and adapting to the game and changes better.**

Up until now, in o!m when facing these issues our only options are to cheat the skin, which often results in floating notes everywhere except on the receptors, or being forced to play on some scrolls speeds that maybe we'd like to change without the complexity of doing so and changing one's habits without the QoL of the visual offset.

Regarding how to integrate it, into the gameplay, I can't say I know how.
However, it has been done in a lot of other VSRGs without any issue and brilliant success. (Even with hit sounds)

Regarding the hit sounds however, it is important to note that they are overall a pretty bad thing to have turned on in VSRGs, and it's pretty damn hard to find good players actually having hit sounds.
It defies rhythm, musical enjoyability and it's even worse when the map isn't properly mapped.

One solution that comes to mind would be that as soon as you move the visual offset, a small prompt says "careful, modifying visual offset beyond it's base value will deactivate hit sounds".

Hoping this gets reads and that people realize how good of a tool this can be,
Kindly,
Waza.

**Reply by Skubers** [reply]
Tokaku actually created a pretty in-depth video some time ago about different offset types in different games (osu! included) and how different types of offsets can be beneficial in different scenarios
https://www.youtube.com/watch?v=JWj98KJVuAQ

**Comment by Paturages** [comment]
osu!mania players, myself especially included, have been working around the lack of visual offset in osu! through modifying our skins in unconventional ways. I personally need to have receptors higher than my hit position to compensate the way I read notes.

In order to demonstrate the practical utility of a visual offset in addition to an audio one, I am providing a series of experimental videos showcasing my live osu!mania gameplay with different visual offset settings, aided by skinning/hit position tinkering and OBS.

For the purposes of the following demonstration videos, I will be operating with the following receptor setup:
![image](https://github.com/ppy/osu/assets/9801350/c296a5d7-bd99-4709-ac89-d059bbcc31f1)

Notice how the arrow receptor is pretty high up: this will allow to demonstrate more explicitly different hit position configurations.

[Link to the skin used in "Visual offset" demonstrations (this is a .zip file: rename to .osk and use in stable)](https://github.com/ppy/osu/files/13798741/paturages.rei.nilou.offset.test.zip)  
[Skin used in "Global offset" (my usual stable skin: receptors are lower, hitpos is below receptors)](https://github.com/ppy/osu/files/13798828/paturages.rei.nilou.zip)

[Beatmap link for "Visual offset" tests](https://osu.ppy.sh/beatmapsets/1138425#mania/2597343)  
[Beatmap link for "Global (audio) offset"](https://osu.ppy.sh/beatmapsets/856084#mania/1788654)

All scores are done on [Stable 20231219.2](https://osu.ppy.sh/home/changelog/stable40/20231219.2) on ScoreV2 for more detailed accuracy feedback.

## Visual offset 1 (no receptor offset)

[Video 1](https://www.youtube.com/watch?v=3F9_WOJ7NAI)

- Audio setup: Game audio on speakers, loud enough to be heard along with the keytaps. Game audio is muted in footage, in favor of microphone audio.
- Audio offset (stable "global offset"): -62 ms
- Hitsound volume: 0%
- Skin: Regular (arrows fall on receptors, hit position = 365)

I am reading visually from an OBS windowed projector, as per shown in the footage

Run 1: Regular setup (no visual offset)
- Reference footage
- You can hear keytaps and music being synced, which confirms the audio offset

Run 2: Add +50ms "Render Delay" filter in OBS
- This simulates monitor delay
- Reading visually offsets the keytaps and the music
- Reading visually results in hits being late in-game

Run 3: Move the receptors below the screen boundaries
- This emulates reading higher than usual, over-anticipating hits
- Reading visually offsets the keytaps and the music
- Reading visually results in hits being early in-game

## Visual offset 2 (positive receptor offset)

[Video 2](https://www.youtube.com/watch?v=5ltTVx5Sz-Y)

- Audio setup: Game audio on speakers, loud enough to be heard along with the keytaps. Game audio is muted in footage, in favor of microphone audio.
- Audio offset (stable "global offset"): -62 ms
- Hitsound volume: 0%
- Skin: Arrows fall below receptors (positive visual offset, hit position = 390)

I am reading visually from an OBS windowed projector, as per shown in the footage

Run 1: Regular setup (no visual offset)
- Reading visually according to the receptors results in early hits
- Hitting early results in the notes actually landing on the receptors

Run 2: Add +50ms "Render Delay" filter in OBS
- This compensates the effect of the receptor offset
- Hitting on time results in the notes landing below the receptors

Run 3: Move the receptors below the screen boundaries
- This furthers the earlyness of the hits

## Visual offset 3 (negative receptor offset)

[Video 3](https://www.youtube.com/watch?v=O_4-QXmP7o8)

- Audio setup: Game audio on speakers, loud enough to be heard along with the keytaps. Game audio is muted in footage, in favor of microphone audio.
- Audio offset (stable "global offset"): -62 ms
- Hitsound volume: 0%
- Skin: Arrows fall above receptors (negative visual offset, hit position = 340)

I am reading visually from an OBS windowed projector, as per shown in the footage

Run 1: Regular setup (no visual offset)
- Reading visually according to the receptors results in late hits
- Hitting late results in the notes actually landing on the receptors

Run 2: Add +50ms "Render Delay" filter in OBS
- This furthers the lateness of the hits

Run 3: Move the receptors below the screen boundaries
- This compensates the effect of the receptor offset
- By moving the receptors below the screen, we can imagine the receptors where they are supposed to be
- Hitting on time results in the notes landing above the receptors

# Global (audio) offset

[Video 4](https://www.youtube.com/watch?v=CXkPTl7_SQY)

- Audio setup: Speakers are **muted** on my end and I'm playing **without sound**, only visually. I routed my audio to another device and game audio is toggled on in resulting video footage for hitsound demonstration.
- Hitsound volume: 100%
- Skin: Arrows fall below receptors (sorry, I can't play without sound accurately in any other way, I need visual offset in most other rhythm games)

This also serves as a testament that I am in fact able to play purely by visuals, as all of those plays are done live and back to back with no adjustment besides offset whatsoever.

Run 1: +0 ms global offset
- Reference run: keytaps and game audio are synced together
- Hitsounds follow along

Run 2: -62 ms global offset (offset confirmed to be accurate to keytaps from visual offset testing)
- Keytaps come later than game audio in the footage as a result of the global offset
- Hitsounds are following the keytaps (i.e. they come late)

Run 3: +62 ms global offset
- Keytaps come earlier than game audio in the footage as a result of the global offset
- Hitsounds are following the keytaps (i.e. they come early)

[Bonus run: +300 ms global offset](https://www.youtube.com/watch?v=sFHl8e-Dw9o)
- Extreme example to make Run 3 more explicit
- I had to restart twice because "first note lag" as a result of high global offset makes it difficult to get good accuracy
- Keytaps come earlier than game audio in the footage as a result of the global offset
- Hitsounds are following the keytaps (i.e. they come early)

-----

I hope this helps provide more concrete references and practical examples in regards to discussing potential solutions and implementations. Let me know if you need more information.

**Reply by OWazabi** [reply]
That is such a detailed answer.
Thanks for putting in the work and effort onto this, it is really important this is taken seriously, literally game defining.

**Comment by Zyfarok** [comment]
I'm mostly a std player but I played mania a bit.

I totally see why people would want to have a visual offset and the reasons have been well exposed by the people above. My understanding is that mania players often play without hitsounds as maps are rarely made with them in mind (and delayed hitsounds can hurt more than it helps if you have high audio latency, especially on fast paced maps). So while the osu! players compensate the audio delay intuitively based on hitsounds delay, the mania players can not and will instead subconsciously time based on keyboard noise/touch (which has 0 delay) and thus they will compensate the audio delay through audio offset. Then, since every player has a different audio and visual delay, the position of notes on the screen might be weird, so the player will want to also apply visual offset. Most of it can be achieved via skinning (that is, moving notes up or down in the sprite) but it is very annoying to do.
One question might be if this offset should also affect hit animations (which can only happen if delay is increased though... unless objects are forced to disappear before the hit) or other things that can not be done through skinning ?

However, one VERY IMPORTANT POINT I want to make is that this visual offset, if introduced, should probably be ruleset specific and if there is any in the osu! ruleset, it SHOULD NOT change anything that couldn't already be done via skinning as it would otherwise potentially be completely broken when used make high AR and low AR much easier. (that is, the setting could apply a scaling to approach-circles, but shouldn't do anything else, like changing apparition timing) This is not an issue in mania because there is no such thing as AR and players can choose their scroll speed, but the same concern probably applies to taiko and catch.

**Reply by Zyfarok** [reply]
@ominoussage
First, if you actually want to time on the physical keyboard hit, then you can change the global audio offset once and for all to compensate your total input to audio latency. So your point seems completely unrelated to visual offset.
But then also, when you play instruments, especially with others, you also learn to deal with latencies (brain loop latency, travel-time (e.g.: sticks), distance (1 meter is 3ms), ...) and can easily learn to compensate your audio latency on osu! too. I played drums for 10 years before starting osu! and I didn't have your problem. (Maybe how much you hear your keyboard has an impact ?)

However, I agree that a high latency can be harder to get used to, and the ~40ms on windows in stable are somewhat high, and that's probably the reason why a lot of top players switched to linux to get <20ms.

**Reply by OWazabi** [reply]
OSU has indeed a tremendous built in delay, at around 40ms indeed.
On top of that you add all the other delay sources, and you get a tremendously delayed audio.
Compare it to any other decent VSRG and you'll basically always find way more delay on OSU.

Playing with hitsounds makes no sense at all, not only it ruins the music and rhythm feeling, but even on a practical pov it is just really bad, and sadly contributes to the fact that many people never learn how to play a rhythm game properly.
Thankfully there's the audio offset, fixing this pretty decently.

The discussion is about visual offset tho, it already makes no doubt that the audio one is required, even tho it doesn't serve half the players that are entirely unable to do anything rhythm related, a true issue if you ask me.
Still, even rhythm players need good visual settings, not to have to force too much on the music sense when the chart gets hard, but mostly on their visual instincts.

Visual Offset is also the BEST tool there will ever be to switch to a new rhythm game, without this setting you could spend a week trying to understand why the fk you can't play properly...
With the visual offset, 5-10 minutes are enough.
We want people to play Lazer, help them do it.
Want people to enjoy and respect OSU as the best rhythm game? Remove all that focus on the hitsounds nonsense.

**Comment by cesarila** [comment]
Just adding my two cents that this is useful to have, for mania especially. I'm visiting my family for a week and brought my laptop with me. Playing with the exact same skin and keyboard that I use on my desktop pc, I was hitting very late even without sound on. I made some skin hacks to try and partially address the issue, but any SV patterns ruin it.

Eventually, I got discouraged and decided to play a different VSRG that has a visual offset setting until I'm back on my desktop PC.

I haven't tried any standard on this setup lately, but I will edit this comment with how that goes if I do give it a shot.

**Reply by Adrriii** [reply]
pretty common experience indeed, having also played with laptops while on trips

**Reply by OWazabi** [reply]
Same thing honestly.
I remember going on a 4 month trip and not understanding at all what the fuck was happening, changing the local offset did nothing for me as I was a beginner back then and couldn't properly play on beat, neither I had the knowledge on how to tweak the computer to make it playable.

However it is been established that OSU doesn't care about this/mania, so the easiest and best solution is to play other actually good VSRGs.
Quaver, Etterna, Interlude... All good games that have proper settings and QoL designed for VSRG.
